src/Univapay/Resources/TransactionToken.php

```diff
@@ -157,6 +157,24 @@ public function deactivate()
         return RequesterUtils::executeDelete($this->getIdContext());
     }
 
+    private function validateCreateCharge()
+    {
+        if ($this->type === TokenType::SUBSCRIPTION()) {
+            throw new UnivapayLogicError(Reason::NON_SUBSCRIPTION_PAYMENT());
+        }
+        if ($this->paymentType === PaymentType::CARD()) {
+            if ($this->data->cvvAuthorize->enabled &&
+                $this->data->cvvAuthorize->status !== CvvAuthorizationStatus::CURRENT()) {
+                throw new UnivapayLogicError(Reason::CVV_AUTHORIZATION_REQUIRED());
+            }
+            
+            if ($this->data->threeDS->enabled &&
+                $this->data->threeDS->status !== ThreeDSStatus::SUCCESSFUL()) {
+                throw new UnivapayLogicError(Reason::THREE_DS_AUTHORIZATION_REQUIRED());
+            }
```

Comment:
This is not necessarily true. A token may have an incomplete or failed 3DS authentication, but still be able to make a charge, similar to a token without any 3DS authentication. They have to do 3DS on every charge if 3DS is mandated.

Reply:
@kgston 🤔 sorry i'm bit lost here, let me clarify this one:

An incomplete or failed token can still be used in the charge/subscription flow, and it will trigger 3DS authentication during the event (i can still follow until this part / correct me if I'm wrong).

> Additionally, add validation to ensure that
> 
> * if `cvv_authorize.enabled = true`, then `cvv_authorize.status == current`
> * if `three_ds.enabled = true`, then `three_ds.status == successful`

My question would be what validation for this one?

Reply:
Ahh, I think I wrote this, it was before this issue was recently brought to my attention.